src/com/dotcms/rest/api/v1/user/UserResourceHelper.java

```diff
@@ -0,0 +1,237 @@
+			SecurityUtils.delayRequest(10, "time-sec");
+			throw new DotDataException("Current user [" + currentUser.getUserId()
+					+ "] does not have the proper 'Login As' role.");
+		}
+		User systemUser = this.userAPI.getSystemUser();
+		final User loginAsUser = this.userAPI.loadUserById(loginAsUserId, systemUser, false);
+		final List<Layout> layouts = this.layoutAPI.loadLayoutsForUser(loginAsUser);
+		if ((layouts == null) || (layouts.size() == 0) || !UtilMethods.isSet(layouts.get(0).getId())) {
+			throw new DotDataException("User [" + loginAsUser.getUserId() + "] does not have any layouts.");
+		}
+		final Role administratorRole = roleAPI.findRoleByFQN(Role.SYSTEM + " --> " + Role.ADMINISTRATOR);
+		if (roleAPI.doesUserHaveRole(loginAsUser, administratorRole)
+				|| roleAPI.doesUserHaveRole(loginAsUser, com.dotmarketing.business.APILocator.getRoleAPI()
+						.loadCMSAdminRole())) {
+			if (!UtilMethods.isSet(loginAsUserPwd)) {
+				throw new DotDataException("The 'Login As' user password is required.");
+			} else if (LoginFactory.passwordMatch(loginAsUserPwd, currentUser) == false) {
+				throw new DotDataException("The 'Login As' user password is invalid.");
+			}
+		}
+		Host host = null;
+		if (UtilMethods.isSet(serverName)) {
+			host = hostWebAPI.findByName(serverName, systemUser, false);
+			if (host == null) {
+				host = hostWebAPI.findByAlias(serverName, systemUser, false);
+			}
+			if (host == null) {
+				host = hostWebAPI.findDefaultHost(systemUser, false);
+			}
+		} else {
+			host = hostWebAPI.findDefaultHost(systemUser, false);
+		}
+		Map<String, Object> sessionData = map(WebKeys.PRINCIPAL_USER_ID, currentUser.getUserId(), WebKeys.USER_ID,
+				loginAsUserId, com.dotmarketing.util.WebKeys.CURRENT_HOST, host);
+		return sessionData;
+	}
+
+	/**
+	 * Gathers all the required parameters that must be changed in order to
+	 * perform the logout process of the 'Login As' functionality.
+	 * 
+	 * @param principalUserId
+	 *            - The ID of the user that performed the 'Login As'
+	 *            functionality and will be restored.
+	 * @param currentLoginAsUser
+	 *            - The temporary {@link User} that was imitated by the
+	 *            principal user.
+	 * @param serverName
+	 *            - The server name obtained via the request URL to try to
+	 *            figure out the Site that will be displayed.
+	 * @return A {@link Map} containing all the parameters that must be changed
+	 *         in order to logout form the 'Login As'.
+	 * @throws DotDataException
+	 *             An error occurred with the input parameters, or when
+	 *             retrieving Site information.
+	 * @throws DotSecurityException
+	 *             The operation is not supported by the user.
+	 */
+	public Map<String, Object> doLogoutAs(final String principalUserId, final User currentLoginAsUser,
+			final String serverName) throws DotDataException, DotSecurityException {
+		User systemUser = this.userAPI.getSystemUser();
```

Comment:
this line could be move after the if validation, since if it is an exception there is not reason to get the System user before.